internal/vm/importers/glob.go

```diff
@@ -0,0 +1,233 @@
+package importers
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"net/url"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/google/go-jsonnet"
+	"github.com/pkg/errors"
+)
+
+const (
+	globParamDirs           = "dirs"
+	globParamVerb           = "verb"
+	globParamStripExtension = "strip-extension"
+)
+
+var supportedParams = map[string]bool{
+	globParamDirs:           true,
+	globParamVerb:           true,
+	globParamStripExtension: true,
+}
+
+var supportedParamsDisplay string
+
+func init() {
+	var ret []string
+	for k := range supportedParams {
+		ret = append(ret, k)
+	}
+	sort.Strings(ret)
+	supportedParamsDisplay = strings.Join(ret, ", ")
+}
+
+// globImportParams are additional parameters for the glob import expressed as URI query params.
+type globImportParams struct {
+	verb           string // one of 'import' or 'importstr'
+	dirs           int    // number of directories levels to retain in the returned object (default: all)
+	stripExtension bool   // whether extensions should be stripped
+}
+
+// keyFor returns the key to be used for the supplied file based on whether directory levels should be limited
+// and/ or extensions stripped.
+func (p globImportParams) keyFor(file string) (string, error) {
+	// short-circuit simple case
+	if p.dirs < 0 && !p.stripExtension {
+		return file, nil
+	}
+	elements := strings.Split(file, string(filepath.Separator))
+	name := elements[len(elements)-1]
+	dirs := elements[0 : len(elements)-1]
+	if p.dirs >= 0 {
+		if len(dirs) > p.dirs {
+			dirs = dirs[len(dirs)-p.dirs:]
+		}
+	}
+	if p.stripExtension {
+		pos := strings.LastIndex(name, ".")
+		if pos > 0 {
+			name = name[:pos]
+		}
+	}
+	finalElements := append(dirs, name)
+	return filepath.Join(finalElements...), nil
+}
+
+func newGlobParams(query url.Values) (params globImportParams, err error) {
+	for k := range query {
+		if !supportedParams[k] {
+			return params, fmt.Errorf("invalid query parameter '%s', allowed values are: %s", k, supportedParamsDisplay)
+		}
+	}
+	verb := query.Get(globParamVerb)
+	if verb == "" {
+		verb = "import"
+	}
+	if verb != "import" && verb != "importstr" {
+		return params, fmt.Errorf("'%s' parameter for glob import must be one of 'import' or 'importstr', found '%s'", globParamVerb, verb)
+	}
+	dirs := -1
+	dirsStr := query.Get(globParamDirs)
+	if dirsStr != "" {
+		l, err := strconv.Atoi(dirsStr)
+		if err != nil {
+			return params, fmt.Errorf("invalid value '%s' for '%s' parameter, %v", dirsStr, globParamDirs, err)
+		}
+		if l < 0 {
+			return params, fmt.Errorf("invalid value '%s' for '%s' parameter, must be non-negative", globParamDirs, dirsStr)
+		}
+		dirs = l
+	}
+	stripExtension := false
+	stripStr := query.Get(globParamStripExtension)
+	if stripStr != "" {
+		switch stripStr {
+		case "false":
+		case "true":
+			stripExtension = true
```

Comment:
```suggestion
			stripExtension = "true" == stripStr
```

Reply:
? That's assigning to `stripExtension`

Reply:
also wants to fail when value is neither `true` nor `false`

Reply:
The value is always true? I am trying to imply that it should be false when query.Get(globParamStripExtension) is "false"

Reply:
go cases do not fall through by default unlike other languages:
https://play.golang.org/p/RRLDNn2IVDZ

Reply:
Damn. I remember using the fallthrough keyword but totally forgot about it.